base/array.jl

```diff
@@ -766,13 +766,25 @@ function collect(itr::Generator)
     et = @default_eltype(itr)
     if isa(isz, SizeUnknown)
         return grow_to!(Vector{et}(), itr)
-    else
+    elseif isa(isz, HasLength)
+        len = length(itr)
+        y = iterate(itr)
+        if y === nothing
+            return et[]
+        end
+        v1, st = y
+        return collect_to_with_first!(Vector{typeof(v1)}(undef, len), v1, itr, st)
+    elseif isa(isz, HasShape)
+        axs = axes(itr)
         y = iterate(itr)
         if y === nothing
-            return _array_for(et, itr.iter, isz)
```

Comment:
Also need to delete the code for `_array_for` just above?

Reply:
Ah, let's keep it then. It is, roughly speaking, an inlined version of this same function, for expanding sufficiently simple (syntax-wise) typed-comprehensions directly to for loops.